8303866: Allow ZipInputStream.readEnd to parse small Zip64 ZIP files

eirbjo opened the pull request:
ZipInputStream.readEnd currently assumes a Zip64 data descriptor if the number of compressed or uncompressed bytes read from the inflater is larger than the Zip64 magic value.

While the ZIP format  mandates that the data descriptor `SHOULD be stored in ZIP64 format (as 8 byte values) when a file's size exceeds 0xFFFFFFFF`, it also states that `ZIP64 format MAY be used regardless of the size of a file`. For such small entries, the above assumption does not hold.

This PR augments ZipInputStream.readEnd to also assume 8-byte sizes if the ZipEntry includes a Zip64 extra information field AND at least one of the 'compressed size' and 'uncompressed size' have the expected Zip64 "magic" value 0xFFFFFFFF. This brings ZipInputStream into alignment with the APPNOTE format spec:

```
When extracting, if the zip64 extended information extra 
field is present for the file the compressed and 
uncompressed sizes will be 8 byte values.
```

While small Zip64 files with 8-byte data descriptors are not commonly found in the wild, it is possible to create one using the Info-ZIP command line `-fd` flag:

`echo hello | zip -fd > hello.zip`

The PR also adds a test verifying that such a small Zip64 file can be parsed by ZipInputStream.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8323583](https://bugs.openjdk.org/browse/JDK-8323583) to be approved

### Issues
 * [JDK-8303866](https://bugs.openjdk.org/browse/JDK-8303866): Allow ZipInputStream.readEnd to parse small Zip64 ZIP files (**Enhancement** - P4)
 * [JDK-8323583](https://bugs.openjdk.org/browse/JDK-8323583): Allow ZipInputStream.readEnd to parse small Zip64 ZIP files (**CSR**)


### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12524/head:pull/12524` \
`$ git checkout pull/12524`

Update a local copy of the PR: \
`$ git checkout pull/12524` \
`$ git pull https://git.openjdk.org/jdk.git pull/12524/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12524`

View PR using the GUI difftool: \
`$ git pr show -t 12524`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12524.diff">https://git.openjdk.org/jdk/pull/12524.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12524#issuecomment-1488264547)

AlanBateman commented:
This proposed change will require a lot of review cycles and a lot of testing. The current implementation came about because of issues with several zip tools so we have to be very careful changing it.

eirbjo commented:
> The current implementation came about because of issues with several zip tools so we have to be very careful changing it.

Thanks for providing historical context, that's very useful!  Do you happen to know a few of the zip tools the current implementation has been tested with?

Let me add that the purpose of the change is mainly to facilitate automated testing of Zip64 features without the need to create ZIP files with sizes exceeding the 0XFFFFFF limit.

(I am not aware of any existing zip tools producing such small Zip64 files, but they would be valid according to the APPNOTE format description)

eirbjo commented:
FWIW, i reverted the refactoring and the CRC == SIGEXT support, leaving only the change to how Zip64 format is determined. Perhaps this may make the change somewhat easier to reason about.

eirbjo commented:
The last version of this PR augments the Zip64 check instead of replacing it, further reducing the risk associated with this proposed change.

eirbjo commented:
With the latest version of this PR, I believe we strictly expand the universe of allowable ZIP entries for ZipInputStream. 

The set of added entries is the set of entries where  a Zip64 extra field exists in the Loc, and where the compressed and uncompressed file data sizes do not exceed 0xFFFFFFFF.

eirbjo commented:
Added even stricter validation to hasZip64Extra checking that the extra data field is right-sized for Zip64.

LanceAndersen commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -87,8 +95,55 @@ public void shouldReadZip64Descriptor() throws IOException {
+        assertEquals(ex.getMessage(), "invalid entry size (expected 0 but got 5 bytes)");
+    }
+
+    private void setExtraSize(short invalidSize) {
```
Please add a comment describing the method (and any others without).  As we add new tests, we are trying to. follow this practice the best that we can.

eirbjo replied:
Added comments for these methods and also for the zip field which I renamed to zip64Zip.

LanceAndersen commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -0,0 +1,154 @@
+
+public class Zip64DataDescriptor {
+
+    private byte[] zip;
```
Please add a comment to the purpose of the field

eirbjo replied:
Comment added.

LanceAndersen reviewed (commented):
Hi Eirik,

Thank you for your continued work in this area.

I have not had a chance to test this yet but will early next week.  Are you aware of any tools that would create this scenario as to the best of my knowledge we have not encountered one that does as of yet?


The comments below are mainly to provide more clarity for future maintainers.

eirbjo commented:
Thanks a lot for looking into this, Lance!

> Are you aware of any tools that would create this scenario as to the best of my knowledge we have not encountered one that does as of yet?

The following on my Mac produces a Zip64 file with an 8-byte data descriptor:

`echo hellohellohellohellohellohellohellohello | zip > hello.zip`

(The Zip64 mode is triggered by the streaming mode, the use of a data descriptor seems sensitive to the length of the input)

eirbjo commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -623,6 +624,24 @@ private void readEnd(ZipEntry e) throws IOException {
+                if (size < 0 || i + size > len) {
+                    break; // Invalid size
+                }
+                i += size + (2 * Short.BYTES);
```
Extracted to a variable with a comment, also replaced the constant 4 earlier in the method.

eirbjo commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -623,6 +624,24 @@ private void readEnd(ZipEntry e) throws IOException {
                 " but got 0x" + Long.toHexString(crc.getValue()) + ")");
         }
     }
+    // Returns true if the ZipEntry has a ZIP64 extended information extra field
```
Added a quite extensive explainer in the Javadocs of this method, including a citation of the relevant APPNOTE.txt spec.

eirbjo commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -575,8 +575,9 @@ private void readEnd(ZipEntry e) throws IOException {
             /* "Data Descriptor" present */
-            if (inf.getBytesWritten() > ZIP64_MAGICVAL ||
-                inf.getBytesRead() > ZIP64_MAGICVAL) {
+            if (hasZip64Extra(e)
```
Added a short comment here, although I'm not fully convinced of the usefulness.

eirbjo commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -0,0 +1,99 @@
+                01000000000000005c000000000000005600000000000000504b06070000
+                0000b20000000000000001000000504b050600000000010001005c000000
+                560000000000""";
+
```
This Zip was produced using my ZIP stream transformation library. This is not in the JDK, so probably not so useful to reference. Let me see if I can rework the test to use a Zip created using the `zip` streaming mode I described instead, that would allow a more independent reproduction.

eirbjo commented:
> Thanks a lot for looking into this, Lance!
> 
> > Are you aware of any tools that would create this scenario as to the best of my knowledge we have not encountered one that does as of yet?
> 
> The following on my Mac produces a Zip64 file with an 8-byte data descriptor:
> 
> `echo hellohellohellohellohellohellohellohello | zip > hello.zip`

Scratch that, I must have looked at the wrong trace output. This prouces a file with a Zip64 extra field, but without the Data Descriptor following the File data.

eirbjo commented:
> > Thanks a lot for looking into this, Lance!
> > > Are you aware of any tools that would create this scenario as to the best of my knowledge we have not encountered one that does as of yet?
> > 
> > 
> > The following on my Mac produces a Zip64 file with an 8-byte data descriptor:
> > `echo hellohellohellohellohellohellohellohello | zip > hello.zip`
> 
> Scratch that, I must have looked at the wrong trace output. This produces a file with a Zip64 extra field, but without the Data Descriptor following the File data.

Ok, it turns out `zip` has a `-fd` option (force descriptor). This allows small Zip64 files with Data Descriptors to be created the following way:

`echo hello | zip -fd > hello.zip`

eirbjo commented:
/issue 8303866

eirbjo commented:
Mark Thomas of Apache chimed in on the JBS to say that this limitation of ZipInputStream is affecting Tomcat migration tooling:

>This bug is affecting the library Apache Tomcat uses to automatically migrate web applications from Java EE to Jakarta EE.
https://github.com/apache/tomcat-jakartaee-migration/issues/46

LanceAndersen commented:
> Mark Thomas of Apache chimed in on the JBS to say that this limitation of ZipInputStream is affecting Tomcat migration tooling:
> 
> > This bug is affecting the library Apache Tomcat uses to automatically migrate web applications from Java EE to Jakarta EE.
> > [apache/tomcat-jakartaee-migration#46](https://github.com/apache/tomcat-jakartaee-migration/issues/46)

It would be useful to obtain the zip/jar in question to validate that your proposed patch addresses the issue as well as verifying if ZipfFile can be used to process the zip/jar as reading the long thread appears that ZipFile:getInputStream is fine.

Do you know how the Zip in question is being created, is it via ApacheCommons and could there be an issue there?

eirbjo commented:
> Do you know how the Zip in question is being created, is it via ApacheCommons and could there be an issue there?

Currently investigating. The Tomcat build is using the Ant `jar` task to create the `tomcat-embed-core.jar`, but then calls `zip` in update mode to update the jar with some Graal related files. This is where the ZIP64 extra fields get introduced.

I'll take a look at testing this PR with this Tomcat jar once I have a better understanding why the `zip` task introduces ZIP64.

LanceAndersen commented:
> > Do you know how the Zip in question is being created, is it via ApacheCommons and could there be an issue there?
> 
> Currently investigating. The Tomcat build is using the Ant `jar` task to create the `tomcat-embed-core.jar`, but then calls `zip` in update mode to update the jar with some Graal related files. This is where the ZIP64 extra fields get introduced.
> 
> I'll take a look at testing this PR with this Tomcat jar once I have a better understanding why the `zip` task introduces ZIP64.

Thank you Eirik, probably need to know which zip and which command line options are being used

eirbjo commented:
> It would be useful to obtain the zip/jar in question to validate that your proposed patch addresses the issue as well as verifying if ZipfFile can be used to process the zip/jar as reading the long thread appears that ZipFile:getInputStream is fine.

The original input file to the Tomcat Migration tool is found here:

https://repo1.maven.org/maven2/org/apache/tomcat/embed/tomcat-embed-core/10.1.7/tomcat-embed-core-10.1.7.jar

I have tested reading this file using `ZipInputStream` in this PR branch and it processes the 8 byte data descriptors successfully.

> Do you know how the Zip in question is being created, is it via ApacheCommons and could there be an issue there?

The file is produced in the Tomcat Ant release targets. First, the `jar` target produces the file, followed by postprocessing by `bnd` for OSGi purposes. At this point the file is still not in ZIP64 mode. Finally, the `zip` task is used to update the jar file, and this is where the ZIP64 seems to be introduced. 

I might want to do some more investigation on the Ant side to see if this behavior could be improved.

eirbjo commented:
/open

eirbjo commented:
Resuscitating this PR since the issue does seem to be causing real problems (see Tomcat Migration tooling discussion in the comments).

Updates:

- `ZipFile` now has `checkZip64ExtraFieldValues`. This method can be reused by `ZipInputStream` when checking the ZIP64 field lengths. To enable reuse, the method is moved to `ZipUtils` and made package protected.

LanceAndersen commented:
We need to hold off this PR until https://github.com/openjdk/jdk/pull/15650 has been integrated as it impacts some of the changes proposed here

eirbjo commented:
> We need to hold off this PR until #15650 has been integrated as it impacts some of the changes proposed here

Thanks Lance, I updated the issue to add [JDK-8314891](https://bugs.openjdk.org/browse/JDK-8314891) as a blocker.

eirbjo commented:
> We need to hold off this PR until #15650 has been integrated as it impacts some of the changes proposed here

Marking this PR ready for review, now that #15650 has been integrated.

A summary of updates after #15650:

- `ZipFile.isZip64ExtBlockSizeValid` cannot be reused from `ZipInputStream` because values from the CEN are not available in streaming mode. My understanding of ZIP64 fields in a LOC is that they should contain csize and/or size, since LOC Offset and Disk Start Number are CEN-only
- Converted the test to junit
- Fixed an incorrect check that a Zip64 block size does not exceed the size of the total extra field

eirbjo commented:
Observation:

Unfortunately, `readLOC` skips reading the LOC's `size`, `csize` and `crc` values when in streaming mode. The fields are also overwritten by `ZipEntry.setExtra0`.

This means we cannot use the original values and compare them to ZIP64_MAGICVAL when determining whether a data descriptor uses 4 or 8 byte values.

LanceAndersen reviewed (commented):
Thanks for tackling this Eirik, I think this is looking good overall.

A few comments below based on my pass through it today.  I want to spend a bit more time looking at ZipInputStream over the coming days.

eirbjo commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -0,0 +1,166 @@
+    @BeforeEach
+    public void setup() {
+        /*
+         * Structure of the ZIP64 file used below . Note the precense
```
Thanks, fixed.

eirbjo commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -0,0 +1,166 @@
+     * @param size the value to set in the 'data size' field.
+     */
+    private void setExtraSize(short size) {
+        int extSizeOffset = 33;
```
Extracted the constant `ZIP64_BLOCK_SIZE_OFFSET` with a comment

eirbjo commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -0,0 +1,166 @@
+         *
+         * The file was produced using the following command:
+         * <pre>echo hello | zip -fd > hello.zip</pre>
+         *
```
Documented the zip command used (zip 3.0, by Info-ZIP), the use of stdin/streaming to enable Zip64, and the use of -fd to force the use of data descriptors.

eirbjo commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -578,7 +578,8 @@ private void readEnd(ZipEntry e) throws IOException {
         if ((flag & 8) == 8) {
             /* "Data Descriptor" present */
-            if (inf.getBytesWritten() > ZIP64_MAGICVAL ||
+            if (hasZip64Extra(e) ||
```
I think I prefer keeping this PR maintaining a strict focus on expanding the set of readable files to include those that use Zip64 extra fields for < 2GB entries with data descriptors.

Would you be ok with that?

Adding validation to `readLOC` is a fair effort, but I would prefer this to be done in a separate PR, similar to your work on adding Zip64 validation to ZipFile.

I wouldn't mind looking into that, but perhaps you would like to handle it, given your comment above about spending some time on `ZipInputStream` in the following days?

LanceAndersen replied:
A follow on PR is fine.  Why don't you take that on seeing you already have your sleeves rolled up in this area :-)

eirbjo commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -627,6 +628,68 @@ private void readEnd(ZipEntry e) throws IOException {
+    private static boolean isZip64ExtBlockSizeValid(int blockSize) {
+        return switch (blockSize) {
+            case 8, 16 -> true;
+            default -> false;
```
So it seems 16 is the only valid size for a Zip64 block in a LOC header. Updated the code to reflect that and also added a comment referencing that original and uncompressed sizes are both required in a LOC header.

LanceAndersen reviewed (approved):
Thank you for your efforts here Eirik.

I am OK with the proposed changes.  We have completed internally our Mach5 runs and thanks to Jai, we have been able to exercise a build against a large internal repository of jars

I would like to suggest that we hold off until we fork for JDK 23 to push this (as we are going to do for my outstanding CEN validation checks) so that we are allowing for more time to catch the unexpected hiccup

eirbjo commented:
Thanks for your patient and thorough review of this long-lived PR, Lance!

No worries, we can hold off the integration until after 22 is forked off mainline.

eirbjo commented:
/integrate

eirbjo commented:
/integrate

eirbjo commented:
/label remove sponsor

jaikiran commented:
Hello Eirik @eirbjo, if this isn't sponsored by tomorrow, then I'll go ahead and do it for you tomorrow. I would have done it now, but there appear to be far too many commits in mainline (540 odd) since this PR was last synced against master branch. Before sponsoring, I'll run a CI test to make sure this doesn't cause any unexpected issues in any area.

AlanBateman commented:
/reviewer 2

eirbjo commented:
/reviewers 2 reviewer

eirbjo commented:
@AlanBateman I was planning to integrate this myself pending the OpenJDK census update for my JDK Commiter status and the Skara team linking my Github account.

But there is no rush here, I'm happy to hold off until you find time to provide feedback. This can wait, please enjoy your holidays! :-)

jaikiran commented:
Hello Eirik, I had a look at this one today. The motivation behind this change appears to be the statement in the specification which says:

```
4.3.9  Data descriptor:

        crc-32                          4 bytes
        compressed size                 4 bytes
        uncompressed size               4 bytes

      4.3.9.1 This descriptor MUST exist if bit 3 of the general
      purpose bit flag is set.
      ...
      For ZIP64(tm) format archives, the compressed and uncompressed 
      sizes are 8 bytes each.

      4.3.9.2 When compressing files, compressed and uncompressed sizes 
      SHOULD be stored in ZIP64 format (as 8 byte values) when a 
      file's size exceeds 0xFFFFFFFF.   However ZIP64 format MAY be 
      used regardless of the size of a file.  When extracting, if 
      the zip64 extended information extra field is present for 
      the file the compressed and uncompressed sizes will be 8
      byte values.  

```
Effectively, this change is proposing to enhance the `java.util.zip.ZipInputStream` to allow for it to parse some more zip/jar files out there, which it currently wouldn't be parsing, even if the spec allowed for such zip files.

Looking through this proposed change, this change only affects `DEFLATED` entries and doesn't impact `STORED` entries. Furthermore, this change affects only those `DEFLATED` entries which have a data descriptor (understandly).

What's being done in this change is that, in a zip/jar file, for each `DEFLATED` entry which has a data descriptor, we now have an additional logic which decides whether we read the compressed/uncompressed sizes in the data descriptor as 4 bytes or as 8 bytes each. Before this change we used to read them as 8 bytes only when the `Inflater` (for that entry) told us that it had dealt with more than `0xFFFFFFFFL` bytes of data for that entry (indicating that this is zip64 entry). With the current proposed change, we not only rely on the `Inflater` for this decision but also rely on `ZipEntry` itself to tell us whether to read 8 bytes or 4 bytes each.

Given this context, I've looked through the changes and I think some additional changes are needed to prevent some potential issues with this proposal. I've added those comments inline.

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -578,7 +578,8 @@ private void readEnd(ZipEntry e) throws IOException {
         if ((flag & 8) == 8) {
             /* "Data Descriptor" present */
-            if (inf.getBytesWritten() > ZIP64_MAGICVAL ||
+            if (hasZip64Extra(e) ||
```
I think it would be better to change the ordering of this boolean conditions. Instead of the current proposed:

```java
if (hasZip64Extra(e) ||
    inf.getBytesWritten() > ZIP64_MAGICVAL ||
    inf.getBytesRead() > ZIP64_MAGICVAL) {

```
the new condition being introduced here - the `hasZip64Extra(e)` should be at the end:

```java
if (inf.getBytesWritten() > ZIP64_MAGICVAL ||
    inf.getBytesRead() > ZIP64_MAGICVAL ||
    hasZip64Extra(e)) {

```

That way the chances of this change running additional code (and thus any unexpected semantic change) for existing zip/jar entries which the Inflater already considers as zip64 is avoided.

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -627,6 +628,70 @@ private void readEnd(ZipEntry e) throws IOException {
+     */
+    private static boolean isZip64ExtBlockSizeValid(int blockSize) {
+        // Uncompressed and compressed size fields are 8 bytes each
+        return blockSize == 16;
```
I'm not following this check. As far as I can see the `blockSize` being passed to this method is the size of the zip64 extra entry and as per the spec:

```
4.5.3 -Zip64 Extended Information Extra Field (0x0001):

      The following is the layout of the zip64 extended 
      information "extra" block. 

      ....

        Value      Size       Description
        -----      ----       -----------
(ZIP64) 0x0001     2 bytes    Tag for this "extra" block type
        Size       2 bytes    Size of this "extra" block
        Original 
        Size       8 bytes    Original uncompressed file size
        Compressed
        Size       8 bytes    Size of compressed data
        Relative Header
        Offset     8 bytes    Offset of local header record
        Disk Start
        Number     4 bytes    Number of the disk on which
                              this file starts 
```
So shouldn't it be 8 + 8 + 8 + 4 = 28 bytes and not 16 bytes? Did I misunderstand the code or the spec?

eirbjo replied:
The size of the "Zip64 Extended Information Extra Field" is determined by the following factors:

- Whether the corresponding field in the LOC or CEN record is larger than the "magic" limit (0xFFFF for "Disk Start Number", 0xFFFFFFFF for "Original Size", "Compressed Size" and "Relative Header Offset"). A component should only appear if the corresponding LOC record field is set to 0xFFFF or 0xFFFFFFFF).
- Whether the Zip64 extra field is contained in a LOC or a CEN header: While Zip64 fields in both headers may have the "Original Size" and "Compressed Size" fields, only the CEN has the corresponding "Relative Header Offset" and "Disk Start Number" fields. So these two components are not relevant in the LOC. (And indeed they are not present in the Info-ZIP test vector used in this PR. See `Zip64DataDescriptor` for the structure)

In this case, we have a Zip64 extra header in a LOC record using "data descriptors". While the spec mandates that LOC headers with the "data descriptor" bit set must have their "size" and "compressed size" fields set to zero, in this case we also have Zip64, so these fields are actually set to 0xFFFFFFFF, and zero is the value used in the Zip64 extra fields. (This means the data size will in this case will always be 16. And we could also add another check verifying that both components are zero as expected when used with a "data descriptor") 

The full text of 4.5.3:

```
4.5.3 -Zip64 Extended Information Extra Field (0x0001):

      The following is the layout of the zip64 extended 
      information "extra" block. If one of the size or
      offset fields in the Local or Central directory
      record is too small to hold the required data,
      a Zip64 extended information record is created.
      The order of the fields in the zip64 extended 
      information record is fixed, but the fields MUST
      only appear if the corresponding Local or Central
      directory record field is set to 0xFFFF or 0xFFFFFFFF.
``` 

It would probably be better to compute the expected size of the Zip64 data by looking at the "size" and "compressed size" in the LOC and add 8 bytes for each field being 0xFFFFFFFF. Then the expected size (8 or 16) can be compared to the actual size. 

I think some recent validation of the Zip64 extra fields in the CEN header was implemented this way. Unfortunately, these values are actually ignored by `readLOC` when it determines the "data descriptor" bit being set, so they are not available to `readEND`.

Perhaps it would be better if we moved the testing for presence of a valid Zip64 extra field into `readLOC` method and have that method set some boolean flag to indicate to readEND whether it should expect 4 or 8 byte fields in the data descriptor? 

This would allow us to check for "size" and "compressed size" being  0xFFFFFFFF AND it would alleviate concerns about trustig the ZipEntry.getExtra value.

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -627,6 +628,70 @@ private void readEnd(ZipEntry e) throws IOException {
+     * This method returns false for any invalid extra block sizes, as if the extra
+     * data contained no Zip64 field.
+     */
+    private boolean hasZip64Extra(ZipEntry e) throws IOException {
```
Given what this method does, I think it shouldn't throw an IOException (or any other exception for that matter), should strictly return true or false. I haven't imported this code locally, but as far as I can see in its implementation, there doesn't seem to be anything that throws a checked IOException, so perhaps this throws is already redundant?

eirbjo replied:
Fixed.

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -627,6 +628,70 @@ private void readEnd(ZipEntry e) throws IOException {
+        int fixedSize = 2 * Short.BYTES; // id + size
+        if (extra != null && extra.length > fixedSize) {
+            for (int i = 0; i < extra.length;) {
+                int id = get16(extra, i);
```
This and other similar calls in this new method have a potential to throw an `ArrayIndexOutOfBoundsException` because we can't trust the `byte[]` returned by the `ZipEntry.getExtra()` call to be actually having the right amount of extra block data. The private `hasMagic` method in  `java.util.jar.JarOutputStream` has an example where we catch the `ArrayIndexOutOfBoundsException` when dealing with the extra data, we should do similar here and return `false`.

(More about the trustworthiness of that byte[] value as a separate comment)

jaikiran commented:
>  With the current proposed change, we not only rely on the Inflater for this decision but also rely on ZipEntry itself to tell us whether to read 8 bytes or 4 bytes each.

The proposed change resides solely in one single internal/private method called `readEnd(ZipEntry e)` of the `java.util.zip.ZipInputStream`. This method gets called in the internal implementation of the `ZipInputStream` when the contents of a "current" `ZipEntry` are being read and this `readEnd()` method gets passed that "current" `ZipEntry`. The passed `ZipEntry`'s `getExtra()` method is then called by the new proposed change to get hold of the byte[] representing the extra blocks of the zip entry and that byte[] is parsed to decide whether we need to read 8 bytes for compressed/uncompressed fields in the data descriptor. It's important to note that before this proposed change, this `readEnd()` method doesn't read anything out of the passed `ZipEntry` and instead only updates it, so before this change, the `ZipEntry` itself doesn't play a role in what data this `readEnd()` method processes.

The crucial part here is that the "current" `ZipEntry` is the one which was created by a previous call to `ZipInputStream.getNextEntry()` method, which is a public overridable method. Furthermore, the `ZipEntry.getExtra()` method too is overridable. So in theory, the byte[] data returned by the `ZipEntry.getExtra()` isn't controlled or validated and can be anything that the overridden method might return. So at a minimum, the newly introduced `hasZip64Extra()` private method which parses this byte[] shouldn't expect this to be parsable and thus should handle any exceptions this might generate and thus the recommendation of catching those exceptions and returning `false`.

The bigger question however is, should `readEnd()` now be entrusted the responsibility of reading the `ZipEntry.getExtra()` byte[] contents that can be anything and even if the content of the byte[] seem to represent a zip64 entry (i.e. the newly introduced `hasZip64Extra()` method returning `true`), should that decision be trusted to further lead the `readEnd()` method to read the compressed/uncompressed sizes in data descriptor as 8 bytes? Would this need additional checks of some form?

Alan and Lance have more experience in this area and I would prefer hearing their thoughts.

jddarcy commented:
/csr needed

jddarcy commented:
Sounds like a CSR is needed for the behavioral change proposed here.

eirbjo commented:
> The crucial part here is that the "current" `ZipEntry` is the one which was created by a previous call to `ZipInputStream.getNextEntry()` method, which is a public overridable method. Furthermore, the `ZipEntry.getExtra()` method too is overridable. So in theory, the byte[] data returned by the `ZipEntry.getExtra()` isn't controlled or validated and can be anything that the overridden method might return. 

Yes, we cannot rely on the extra field contents to be well-formed and valid. Because of this, `hasZip64Extra` must be coded in a defensive fashion, meaning it returns true only when it finds a valid Zip64 extra field, false otherwise. It should also never throw exceptions. This is the case regardless of ZipEntry concerns, since the input ZIP data may already be malformed.

> The bigger question however is, should `readEnd()` now be entrusted the responsibility of reading the `ZipEntry.getExtra()` byte[] contents that can be anything and even if the content of the byte[] seem to represent a zip64 entry (i.e. the newly introduced `hasZip64Extra()` method returning `true`), should that decision be trusted to further lead the `readEnd()` method to read the compressed/uncompressed sizes in data descriptor as 8 bytes? Would this need additional checks of some form?

To avoid readEND relying on "untrusted" ZipEntry contents, I moved the call to `hasZip64Extra` from `readEND` to `readLOC`. This allows `hasZip64Extra` to work on the raw extra data.

 For this to work, I needed to add a new internal boolean field `ZipInputStream.expectEightBitDataDescriptor`. When true, this flag indicates to `readEND` that the next data descriptor has 8-byte size fields.

eirbjo commented:
Putting this PR in draft mode while I figure out how to address review comments.

eirbjo commented:
Marking this PR ready for review again with the following changes applied:

- The decision of whether to expect 64 bit data descriptors is moved from `readEnd` to `readLOC`. This allows access to the 'compressed size' and 'uncompressed size' of the LOC as well as direct access to the extra data, remediating concerns raised by @jaikiran about trusting any passed `ZipEntry` extra data.
- Checking of LOC is tightened to require that the 'compressed size' and 'uncompressed size' fields are both set to the Zip64 magic marker 0xFFFFFFFF (Required for Zip64 entries)
- Checking of the Zip64 field is tightened to require that the 'Original Size' and 'Compressed Size' are both present and set to zero. (Required for Zip64 entires using Data Descriptors)
- An new internal boolean field `ZipInputStream.expect64BitDataDescriptor` passes the decision made in `readLOC` on to the `readEnd` method.

The two support methods has been renamed:

- `hasZip64Extra` is now called `expect64BitDataDescriptor`
- 'isZip64ExtBlockSizeValid` is now called `isZip64DataDescriptorField`

eirbjo commented:
> Sounds like a CSR is needed for the behavioral change proposed here.

Thanks for flagging this @jddarcy 

I'm personally not 100% convinced a CSR is warranted in this case, but I'm of course happy to extend the following into a CSR if you or other reviewer think that's the best way forward:  

Let's review the intended and unintended behavioral changes in this PR:  

The _intended_ behavioral change is to extend the set of ZIP entries parsable by `ZipInputStream` to include "small Zip64 entries". Such entries meet the following criteria:
 
1. They are clearly marked as using the Zip64 format, meaning that the LOC's 'compressed size' and 'uncompressed size' fields are set to the "Zip64 magic value" 0xFFFFFFFF and that the LOC's extra field includes a valid _Zip64 Extended Information Field_.
2. They use _streaming mode_, meaning that the 'general purpose bit flag' 3 is set, that the Zip64 'Original Size' and 'Compressed Size' are both set to zero and that file data is followed by a 'Data Descriptor' containing the actual size values.
3. The Data Descriptor represents the 'compressed size' and 'uncompressed' size fields using 8 byte fields, instead of the regular 4 byte fields.

These entires are currently not parsable by `ZipInputStream`. Trying to parse them fails with an exception similar to the following:

```
java.util.zip.ZipException: invalid entry size (expected 0 but got 6 bytes)
	at java.base/java.util.zip.ZipInputStream.readEnd(ZipInputStream.java:616)
```

The _unintended_ behavioral change in this PR is that ZIP entries marked as Zip64, but still having a Data Descriptor using 4 bytes to represent the size fields will now become unparsable. Such entries meet critera 1 and 2 above, but not 3.

While I have not performed a corpus search, I believe such entries are rare, if they exist at all:

- They would be in clear violation of the `APPNOTE.txt` spec, which in 4.3.9.2 says  _ZIP64 format MAY be used regardless of the size of a file.  When extracting, if the zip64 extended information extra field is present for the file the compressed and uncompressed sizes will be 8 byte values_
- The `zipdetails` tool fails to parse such a file, instead reading the two 4-byte numbers into one 8-byte 'compressed size'
- The file cannot be parsed by external APIs similar to `ZipInputStream`, such as the Python `stream-unzip` module.

To avoid unintended behavioral changes in the implementation, care has been taken to make the Zip64 extra field validation code robust to invalid input, including out of bounds array access. 

To summarize, this PR aims to allow a set of useful, valid ZIP files to be parsed, at the cost of potentially making a small set of invalid ZIP files unparsable.

jddarcy commented:
> > Sounds like a CSR is needed for the behavioral change proposed here.
> 
> Thanks for flagging this @jddarcy
> 
> I'm personally not 100% convinced a CSR is warranted in this case, but I'm of course happy to extend the following into a CSR if you or other reviewer think that's the best way forward:
> 
> Let's review the intended and unintended behavioral changes in this PR:
> 
> The _intended_ behavioral change is to extend the set of ZIP entries parsable by `ZipInputStream` to include "small Zip64 entries". Such entries meet the following criteria:
> 
>     1. They are clearly marked as using the Zip64 format, meaning that the LOC's 'compressed size' and 'uncompressed size' fields are set to the "Zip64 magic value" 0xFFFFFFFF and that the LOC's extra field includes a valid _Zip64 Extended Information Field_.
> 
>     2. They use _streaming mode_, meaning that the 'general purpose bit flag' 3 is set, that the Zip64 'Original Size' and 'Compressed Size' are both set to zero and that file data is followed by a 'Data Descriptor' containing the actual size values.
> 
>     3. The Data Descriptor represents the 'compressed size' and 'uncompressed' size fields using 8 byte fields, instead of the regular 4 byte fields.
> 
> 
> These entires are currently not parsable by `ZipInputStream`. Trying to parse them fails with an exception similar to the following:
> 
> ```
> java.util.zip.ZipException: invalid entry size (expected 0 but got 6 bytes)
> 	at java.base/java.util.zip.ZipInputStream.readEnd(ZipInputStream.java:616)
> ```
> 
> The _unintended_ behavioral change in this PR is that ZIP entries marked as Zip64, but still having a Data Descriptor using 4 bytes to represent the size fields will now become unparsable. Such entries meet critera 1 and 2 above, but not 3.
> 
> While I have not performed a corpus search, I believe such entries are rare, if they exist at all:
> 
>     * They would be in clear violation of the `APPNOTE.txt` spec, which in 4.3.9.2 says  _ZIP64 format MAY be used regardless of the size of a file.  When extracting, if the zip64 extended information extra field is present for the file the compressed and uncompressed sizes will be 8 byte values_
> 
>     * The `zipdetails` tool fails to parse such a file, instead reading the two 4-byte numbers into one 8-byte 'compressed size'
> 
>     * The file cannot be parsed by external APIs similar to `ZipInputStream`, such as the Python `stream-unzip` module.
> 
> 
> To avoid unintended behavioral changes in the implementation, care has been taken to make the Zip64 extra field validation code robust to invalid input, including out of bounds array access.
> 
> To summarize, this PR aims to allow a set of useful, valid ZIP files to be parsed, at the cost of potentially making a small set of invalid ZIP files unparsable.

@eirbjo, either of the intended or unintended effects of the PR on their own would merit a CSR and this is a case where the inclusive-OR applies so a CSR is definitely needed; thanks.

eirbjo commented:
> @eirbjo, either of the intended or unintended effects of the PR on their own would merit a CSR and this is a case where the inclusive-OR applies so a CSR is definitely needed; thanks.

Thanks Joe!

 A CSR has been proposed and is ready for the first round of feedback: https://bugs.openjdk.org/browse/JDK-8323583

AlanBateman commented:
I was initially concerned about this change but I think it's okay. As you say, it expands set the set of ZIP file that can be read by ZipInputStream. I've added myself as reviewer to the CSR.

turbanoff commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -0,0 +1,243 @@
+        // Create the ZIP file used for testing that invalid Zip64 extra fields are ignored
+        // This ZIP has the regular 4-bit data descriptor
+
+        byte[] extra = new byte[Long.BYTES  + Long.BYTES + Short.BYTES * 2]; // Size of a regular Zip64 extra field
```
```suggestion
        byte[] extra = new byte[Long.BYTES + Long.BYTES + Short.BYTES * 2]; // Size of a regular Zip64 extra field
```

turbanoff commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -627,6 +640,91 @@ private void readEnd(ZipEntry e) throws IOException {
+     * @param csize the value of the 'compressed size' field in the LOC
+     * @param size  the value of the 'uncompressed size' field in the LOC
+     */
+    private boolean expect64BitDataDescriptor(byte[] extra, int flag, long csize, long size)  {
```
```suggestion
    private boolean expect64BitDataDescriptor(byte[] extra, int flag, long csize, long size) {
```

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -523,6 +526,13 @@ private ZipEntry readLOC() throws IOException {
+        expect64BitDataDescriptor = false;
+
+        long csize = get32(tmpbuf, LOCSIZ);
+        long size = get32(tmpbuf, LOCLEN);
```
Hello Eirik, I suspect this part of the change has an issue. Before reading the `tmpbuf` for compressed and uncompressed sizes, there will be 32 bits of CRC, which should be read first. This now skips those 32 CRC bits and reads them (in the else block) after reading these sizes and that can cause incorrect LOC data.

jaikiran replied:
The Github actions job which runs tier1 is all successful with this proposed change. So I'm a bit surprised that the tests didn't catch any issues, which makes me wonder if we have enough test coverage that covers this change.

eirbjo replied:
> Hello Eirik, I suspect this part of the change has an issue. Before reading the `tmpbuf` for compressed and uncompressed sizes, there will be 32 bits of CRC, which should be read first. This now skips those 32 CRC bits and reads them (in the else block) after reading these sizes and that can cause incorrect LOC data.

How does the order of the reads from the byte array matter? Outside cache efficiency I would presume they could be read in any order?

(These reads are from a temp buffer, not the stream)

Could you elaborate? I'm not sure I'm following :-)

jaikiran replied:
>(These reads are from a temp buffer, not the stream)

Ah! you are right indeed. I didn't correctly read that part of that code. It's reading from a temp buffer which has been fully initialized with a LOC and these reads happens with specific offsets within that buffer. So yes, you are correct that the order won't matter here. Thank you for that detail.

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -627,6 +640,91 @@ private void readEnd(ZipEntry e) throws IOException {
+        }
+
+        // The LOC's 'compressed size' and 'uncompressed size' must both be marked for Zip64
+        if (csize != ZIP64_MAGICVAL || size != ZIP64_MAGICVAL) {
```
The spec for this says different. It says:

>
> 4.4.4 general purpose bit flag:
> ...
>    Bit 3: If this bit is set, the fields crc-32, compressed size and uncompressed size are set to zero in the local header.  The correct values are put in the data descriptor immediately following the compressed data.  

So it expects the value zero for the compressed/uncompressed sizes in the LOC when the data descriptor bit is set.

eirbjo replied:
The spec isn't terribly helpful in spelling out what should happen in the case where an entry combines the uses of  data descriptors (mandating that CRC, size and compressed size must be zero) with Zip64 (mandating that size and compressed size must be 0xFFFFFFFF)

My interpretation (based in the InfoZIP implementation) is that in such cases, CRC should be zero, while size and compressed size should be 0xFFFFFFFF, with their counterparts in the Zip64 extra field should set to zero.

Does this make sense?

jaikiran replied:
> The spec isn't terribly helpful in spelling out what should happen in the case where an entry combines the uses of data descriptors (mandating that CRC, size and compressed size must be zero) with Zip64 (mandating that size and compressed size must be 0xFFFFFFFF)

I see what you mean. However, given that the data descriptor general bit field is set to indicate that the data descriptor should be honoured, plus the spec stating that the original/compressed sizes in the zip64 extra block aren't mandatory, I think not relying on the zip64 extra block for parsing decisions about a data descriptor content might be better. I think the only role that a zip64 block should play when we are deciding how to parse a data descriptor is whether or not the entry has zip64 extra field set (the header id value of the extra field). Does that sound reasonable?

eirbjo replied:
> I think the only role that a zip64 block should play when we are deciding how to parse a data descriptor is whether or not the entry has zip64 extra field set (the header id value of the extra field). Does that sound reasonable?

Are you suggesting that we ONLY look for the presence of a Zip64 (tag 0x1) extended field? Meaning we should ignore the following:

- The values of the 'compressed size' and 'uncompressed size' fields in the LOC header
- The contents/fields in the data block in the Zip64 header

The purpose of this check is to determine wheter the entry "is in the ZIP64 format", to use parlance from `APPNOTE.TXT`. The reason stronger validation was added was to avoid false positives, meaning an entry would be interpreted as "in the Zip64 format", even when that was not the intention of the producer. Some of this validation was on my initiative, but I think maybe some was in response to concerns raised in review.

If we ignore the 'uncompressed size' and 'compresssed size' fields, then if a LOC has both of these set to zero and has a Zip64 extended field, then we'll intepret this as "in the Zip64 format", even though no field is marked using the Zip64 magic value `0xFFFFFFFF`.

I think I would be fine with dropping inspection of the Zip64 contents (as there are probably efforts underway to add validation of LOC Zip64 fields similar to the recently added CEN validation).

But I'm a bit worried that ignoring the LOC values will make us parse some data descriptors using 8 bytes, when the producer did in fact not intend to use the Zip64 format. (The Zip64 field just came through some kind of "pollution" or unintended copying.

A natural way to implement Zip64 parsing is to only look at the extended field if at least one of the relevant LOC headers are `0xFFFFFFFF`, otherwise one can short-circuit and assume there is no Zip64 field.

@LanceAndersen Do you have a cent or two to spare?

LanceAndersen replied:
> @LanceAndersen Do you have a cent or two to spare?

Let me try and dig out from a couple of things and circle back to this again

jaikiran commented on src/java.base/share/classes/java/util/zip/ZipInputStream.java:
```diff
@@ -627,6 +640,91 @@ private void readEnd(ZipEntry e) throws IOException {
+     * @param blockSize the size of the extra field data block
+     * @return true if the extra field is a Zip64 extra field compatible with data descriptors
+     */
+    private static boolean isZip64DataDescriptorField(int headerId, byte[] extra, int blockStart, int blockSize) {
```
I understand the goals of this method - what it's trying to do is, assure the caller that the extra field/block actually is a zip64 extra block. That assurance is then used to access the data descriptor content as 8 byte fields. 
However, I think in this proposed implementation of this method we are perhaps doing a bit too much. Specifically, I don't think we should check what values have been stamped for "Original size" and "Compressed size" fields of this zip64 block. I think, those values (presence or absence) shouldn't play a role in deciding whether we have to read a data descriptor size fields as 8 bytes. Doing these checks for these zip64 original/compressed size fields, I think will open up more permutations about which zip entries get processed as 8 byte data descriptors.

Given the context in which this method is used, I think the only checks that we should do in this method is to verify that the header id is `ZIP64_EXTID`. 
Perhaps then this `isZip64DataDescriptorField(...)` won't be needed and we can just inline that `headerid == ZIP64_EXTID` check inline in the implementation of `expect64BitDataDescriptor(...)` method

jaikiran commented:
On a general note - thank you for updating this PR from the previous proposed approach. The current proposed approach of solely relying on the data that comes from within the stream to decide whether or not to use 8 bytes for a data descriptor compressed/uncompressed fields, looks right to me. That prevents issues related to basing this decision on some application controlled/manipulated data which may not match the stream content.

eirbjo commented:
In help make progress here, I have relaxed the validation here such that we now check:

- That the "streaming mode" bit 3 flag is set
- That at least one of the LOC's size fields are marked 0xFFFFFFFF.
- That the LOC extra field has a field with header ID 0x1 (Zip64)

Any reading/validation of the contents of the Zip64 extra field has been removed.

@jaikiran Is this closer to what you'd like to see?

LanceAndersen commented:
> To help make progress here, I have relaxed the validation such that we now check:
> 
> * That the "streaming mode" bit 3 flag is set
> * That at least one of the LOC's size fields are marked 0xFFFFFFFF.
> * That the LOC extra field has a field with header ID 0x1 (Zip64)
> 
> Any reading/validation of the contents of the Zip64 extra field has been removed.
> 
> @jaikiran Is this closer to what you'd like to see?

Have not forgotten this, hope to get to it next week

jaikiran reviewed (approved):
Hello Eirik, thank you for the latest round of updates. These latest changes in commit `4af7f500` look good to me. I just have a trivial suggestion for the test which I've added inline.

jaikiran commented on test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:
```diff
@@ -0,0 +1,275 @@
+    private void readZipInputStream(byte[] zip) throws IOException {
+        try (ZipInputStream in = new ZipInputStream(new ByteArrayInputStream(zip))) {
+            ZipEntry e;
+            while ( (e = in.getNextEntry()) != null) {
```
The zip is expected to have a single ZipEntry. Would it be better to do something like this, so that if the ZipEntry is missing, then the test fails?

(I haven't tested this code below)

```java
ZipEntry e = in.getNextEntry();
assertNotNull(e, "missing zip entry");
assertEquals("hello\n", new String(in.readAllBytes(), StandardCharsets.UTF_8));
assertNull(in.getNextEntry(), "unexpected additional zip entry");

```

jaikiran reviewed (approved):
Thank you for the update, looks good to me.

LanceAndersen reviewed (approved):
The latest updates seem OK.

Thank you Eirik

eirbjo commented:
Thanks @LanceAndersen, @jaikiran and @AlanBateman for your helpful and patient guidance.

/integrate
